### PR TITLE
Initial Infrastructure for GeneVC revamp

### PR DESCRIPTION
**Pull request by sarahscott**
- [x] Made a view model

Currently thinking about the networking components and if I should leave them as-is or abstract them in another way. The view model will soon be a useful place to put refinement logic - I know it's pretty sparse right now.

### CODE REVIEW

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">1 Error</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:no_entry_sign:</td>
      <td>There were <a href="https://eigen-ci.s3.amazonaws.com/2263/index.html">snapshot errors</a></td>
    </tr>
    
  </tbody>
</table>
  

  
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">2 Warnings</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td>PR is classed as Work in Progress</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td><h4>Found unstubbbed networking requests</h4>

<ul>
<li><a href="https://github.com/search?q=ARFairFavoritesNetworkModel+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARFairFavoritesNetworkModelSpec</a> in <code>test_ignores_artworks_without_a_partner</code></li>
<li><a href="https://github.com/search?q=ARFairFavoritesNetworkModel+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARFairFavoritesNetworkModelSpec</a> in <code>test_ignores_artworks_without_a_partner</code></li>
<li><a href="https://github.com/search?q=ARFairFavoritesNetworkModel+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARFairFavoritesNetworkModelSpec</a> in <code>test_ignores_artworks_without_a_partner</code></li>
</ul>
</td>
    </tr>
    
  </tbody>
</table>
  

  


<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

File: Artsy/Views/Collection_Views/GeneViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+
+extension GeneViewModel {
+    func geneHasDescription() -> Bool {
+        return gene.geneDescription.characters.count > 0
+    }
+
+    func displayName() -> String {
+        return gene.name
+    }
```
Comment (orta) [inline Artsy/Views/Collection_Views/GeneViewModel.swift]: these would feel better as computed properties, https://github.com/artsy/eigen/blob/master/Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionLotViewModel.swift#L75

Comment (sarahscott) [inline Artsy/Views/Collection_Views/GeneViewModel.swift]: I like that idea

**Comment by sarahscott** [comment]
Right now, I'm trying to make the view controller really basic and strictly concerned with UI tweaking, the network model the sole source of all network requested material, and the view model for presentation logic. Any objections to this pattern?

**Comment by orta** [comment]
Nah, that makes sense :+1: , but do consider that you don't have to reinvent the wheel too much, a lot of the AuctionViewController contains generic patterns that should be transferrable ( and if they're not then that's worrying )

**Comment by sarahscott** [comment]
I'm taking a lot of inspiration from the AuctionVC. It's interesting to port the objc ``get:success failure:failure`` pattern to swift; is there a nicer way to do that than what I've done so far?

File: Artsy/Views/Styled_Subclasses/Buttons/ARNavigationButton.m [hunk 2]
```diff
@@ -194,7 +194,6 @@ - (id)initWithFrame:(CGRect)frame withBorder:(CGFloat)borderWidth
     }
 
     [self setNeedsUpdateConstraints];
-    [self updateConstraintsIfNeeded];
```
Comment (sarahscott) [inline Artsy/Views/Styled_Subclasses/Buttons/ARNavigationButton.m]: I'm going to remove this for now so the CI snapshot errors leave me alone

File: Artsy/Views/Collection_Views/GeneViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    var sharingController: ARSharingController {
+        return ARSharingController.init(object: gene, thumbnailImageURL: gene.smallImageURL())
+    }
+    
+    var analyticsDictionary: NSDictionary? {
+        guard gene == gene else { return nil }
+        return [ "gene" : gene.geneID, "type" : "gene" ]
+    }
+    
+    // for spotlight; could use some feedback on how best to deal with this one
```
Comment (sarahscott) [inline Artsy/Views/Collection_Views/GeneViewModel.swift]: :rotating_light:

**Comment by orta** [comment]
This should probably be a new UIViewController that you can then replace the ARGeneVC with once it's done so that you can keep pushing PRs without breaking builds, get it on a feature flag in ARRouter :+1:

**Comment by orta** [comment]
Hrm, the scope of this PR looks to be basically "Build a new GeneVC" which I'm not really sure I want to be working on without incremental PRs, will try to find a way to cut the scope of this down to size and get it covered in tests etc

**Comment by sarahscott** [comment]
Cool cool. If it isn't useful anymore we can also just delete it #egoless

**Comment by orta** [comment]
haha, nice, it's just grown outside of the scope of "set up the viewmodel + setup routing for new VC" - will get that covered and trim it down

**Comment by orta** [comment]
Alright, @sarahscott this one is back to you for review, ignore that the ModernGeneViewContoller exists WRT tests, I'm going to have that in a separate PR 👍 

I also don't think the GeneViewModel is complicated enough to warrant tests 👍

**Comment by orta** [comment]
and the Danger fail is OK too, the spacing was wrong before

**Comment by sarahscott** [comment]
✨

File: Artsy/Networking/ArtworkSources/ARGeneArtworksNetworkModel.m [hunk 4]
```diff
@@ -25,6 +26,21 @@ - (id)initWithGene:(Gene *)gene
...
+    return [self initWithGene:[[Gene alloc] initWithGeneID:geneID]];
+}
+
+- (void)getGene:(void (^)(GeneViewModel *viewModel))success
+{
+    __weak typeof(self) wself = self;
+    [ArtsyAPI getGeneForGeneID:self.gene.geneID success:^(Gene *gene) {
+        __strong typeof (wself) sself = wself;
+        sself.gene = gene;
+        success([[GeneViewModel alloc] initWithGene:gene]);
```
Comment (alloy) [inline Artsy/Networking/ArtworkSources/ARGeneArtworksNetworkModel.m]: I think you should yield nil here if wself was released.

**Pull request merged**